enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -1252,58 +1253,77 @@ func downloadLayer(ctx context.Context, layer ctr.Layer, destDir string) error {
 	}
 	defer os.RemoveAll(tempUnpackDir)
 
-	// TODO: avoid tar command.
-	cmd := exec.CommandContext(ctx, "tar", "--extract", "--directory", tempUnpackDir)
-	var stderr bytes.Buffer
-	cmd.Stdin = rc
-	cmd.Stderr = &stderr
-	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
-	if err := cmd.Run(); err != nil {
-		return status.UnavailableErrorf("download and extract layer tarball: %s: %q", err, stderr.String())
-	}
-
-	// Convert whiteout files to overlayfs format.
-	err = filepath.WalkDir(tempUnpackDir, func(path string, entry fs.DirEntry, err error) error {
+	tr := tar.NewReader(rc)
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
 		if err != nil {
-			return err
+			return status.UnavailableErrorf("download and extract layer tarball: %s", err)
 		}
 
-		if !entry.Type().IsRegular() {
-			return nil
-		}
+		fileName := strings.TrimPrefix(filepath.Clean(header.Name), "/")
+		target := filepath.Join(tempUnpackDir, fileName)
```

Comment:
That's what I expect `filepath.Clean()` to handle https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/path/filepath/path_test.go;l=54-69;drc=1eee17dc4e796c2fe60e8f9e82b66c00610adc31.

So combining clean + trim prefix + join should get rid of most cases(?) https://go.dev/play/p/6o4loXyQdNv

Do you think we should err out instead? Or we should add a `strings.Contains` check + err after this as a precaution?

Reply:
`Clean` wouldn't help in case `tempUnpackDir` is an absolute path, tar entries could still navigate anywhere on the file system. But even if we check for `..`, this is still possible by creating a symlink first and then using it to jump. 

I think that we should first clarify what the implications of that happening are going to be: is this flexibility a feature, a foot gun or may it have security implications?

Reply:
Playing with the playground a bit more, because we are using `filepath.Join()`, we should not need to apply `TrimPrefix` to the cleaned path to remove leading `/`.

So I removed TrimPrefix as well.

Reply:
> Do you think we should err out instead? 

@sluongng yes, I think we should return an error. If a tar file contains a file name like `../foo` we only have a couple of options: (a) ignore it, (b) return an error, (c) do something incorrect or insecure. I think (b) is the best option because the "ignore" option is sort of failing "silently" which might be confusing imo

Reply:
I added the err to the general case as well as the hardlink case.

I checked for both `strings.Contain` as well as `filepath.Clean` diff so that we can avoid cases like `dir/some..file/`.

Reply:
Isn't `dir/some..file/` totally fine? Only `../`, `/../` and `/..` could be problematic.

Reply:
`/` is also a path separator on Windows. While that's not our focus here, could you convert this with `filepath.ToSlash` and then split on `/`?